File: 0094_Binary_Tree_Inorder_Traversal/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// Recursive T = O(n), S = O(n)
class Solution {
  public:
    vector<int> inorderTraversal(TreeNode *root) {
        if (root == nullptr) {
            return {};
        }
        vector<int> left = inorderTraversal(root->left);
        vector<int> right = inorderTraversal(root->right);
        left.push_back(root->val);
        left.insert(left.end(), right.begin(), right.end());
        return left;
    }
};

// Iteration T = O(n), S = O(n)
class Solution {
  public:
    vector<int> inorderTraversal(TreeNode *root) {
        if (root == nullptr) {
            return {};
        }
        vector<int> res;
        stack<TreeNode *> q;
        while (root != nullptr || !q.empty()) {
            while (root != nullptr) {
                q.push(root);
                root = root->left;
            }
            TreeNode *p = q.top();
            q.pop();
            res.push_back(p->val);
            root = p->right;
        }
        return res;
    }
};

File: 0094_Binary_Tree_Inorder_Traversal/golang/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// Recursive T = O(n), S = O(n)
func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}
	res := []int{}
	dfs(root, &res)
	return res
}

func dfs(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}
	dfs(root.Left, res)
	*res = append(*res, root.Val)
	dfs(root.Right, res)
}

// Iteration T = O(n), S = O(n)
func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}
	st := []*TreeNode{}
	p := root
	res := []int{}
	for len(st) != 0 || p != nil {
		for p != nil {
			st = append(st, p)
			p = p.Left
		}
		n := len(st)
		p = st[n-1]
		st = st[:n-1]
		res = append(res, p.Val)
		p = p.Right
	}
	return res
}